### PR TITLE
Hotfix/2.2.1.1

### PR DESCRIPTION
**Pull request by dals83**

### CODE REVIEW

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 1]
```diff
@@ -55,31 +55,26 @@ def convert(self, item, tmp_dir):
...
-            if err and 'Error' in err.decode('utf-8'):
-                print('pdftohtml convertion error: %s' % err)
-                return False
-        except TimeoutExpired as te:
-            p.kill()
-            # outs, errs = p.communicate()
-            print('Popen command expired: %s' % te)
-            return False
+        p = Popen(command, stdin=PIPE, stdout=DEVNULL, stderr=PIPE)
+        output, err = p.communicate(b"input data that is passed to subprocess' stdin")
```
Comment (magenta-af) [inline scrapy-webscanner/scanners/processors/pdf.py]: (Why _does_ the code randomly type this string into `pdftohtml`, anyway?)

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 2]
```diff
@@ -55,31 +55,26 @@ def convert(self, item, tmp_dir):
...
+
+        if err:
+            print('pdftohtml conversion error: {} \non document {}'.format(err.decode('utf-8'), new_file_path))
 
-        return_code = 0
+        # pdftohtml returns 1 if pdf is not type pdf or if pdf could not be converted.
+        if p.returncode:
```
Comment (magenta-af) [inline scrapy-webscanner/scanners/processors/pdf.py]: This is stylistic pedantry of the highest order, but I really don't like comparing process exit codes in this way -- I think it looks much less magic if you explicitly write `!= 0`.

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 3]
```diff
@@ -55,31 +55,26 @@ def convert(self, item, tmp_dir):
...
         if os.path.exists(result_file):
-            print('resul')
-            return_code = call([
+            check_call([
                 'sed', '-i', 's/\\xff//;s/\\xfe//', result_file
             ])
 
         os.remove(new_file_path)
-        return return_code == 0
+        return p.returncode
```
Comment (danchr) [inline scrapy-webscanner/scanners/processors/pdf.py]: The previous result was a Boolean representing the success value of the method; now it's `False` when failing, and `0` when succeeding — both of which are falsy values. That seems somewhat unconventional, and probably unintentional 😉

```suggestion
        return True
```

File: scrapy-webscanner/scanners/processors/pdf.py (hunk 1 above)
Comment (danchr) [reply scrapy-webscanner/scanners/processors/pdf.py]: > (Why _does_ the code randomly type this string into `pdftohtml`, anyway?)

Good question. This code should probably just do a [`subprocess.run()`](https://docs.python.org/3.5/library/subprocess.html#subprocess.run) instead.

**Comment by dals83** [comment]
> Why not `2.2.2`?
Due to our vejledning:
https://redmine.magenta-aps.dk/projects/dsn/wiki/Vejledning_i_versionsstyring_med_Git_og_Git_Flow

**Comment by danchr** [comment]
>> Why not `2.2.2`?
> Due to our vejledning:
> https://redmine.magenta-aps.dk/projects/dsn/wiki/Vejledning_i_versionsstyring_med_Git_og_Git_Flow

Well, I always thought of the third item as the hotfix one, and that corresponds to the official [Git Flow](https://nvie.com/posts/a-successful-git-branching-model/#creating-the-hotfix-branch) guide 🤷‍♂

**Comment by magenta-af** [comment]
I don't want to block this PR unnecessarily; if you want to go ahead with it and get it released without addressing my comments, then I'm (_more or less_) fine with that.

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 4]
```diff
@@ -55,14 +55,14 @@ def convert(self, item, tmp_dir):
         command.extend(extra_options)
         command.append(new_file_path)
 
-        p = Popen(command, stdin=PIPE, stdout=DEVNULL, stderr=PIPE)
-        output, err = p.communicate(b"input data that is passed to subprocess' stdin")
+        completed_process = run(command, capture_output=True)
```
Comment (magenta-af) [inline scrapy-webscanner/scanners/processors/pdf.py]: `capture_output` will only work [in Python 3.7 and up](https://docs.python.org/3/library/subprocess.html).
> Changed in version 3.7: Added the text parameter, as a more understandable alias of universal_newlines. Added the capture_output parameter.

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 5]
```diff
@@ -55,14 +55,14 @@ def convert(self, item, tmp_dir):
...
+        err = completed_process.stderr
 
-        if err:
+        if err is not None:
             print('pdftohtml conversion error: {} \non document {}'.format(err.decode('utf-8'), new_file_path))
 
         # pdftohtml returns 1 if pdf is not type pdf or if pdf could not be converted.
-        if p.returncode:
+        if completed_process.returncode != 0:
```
Comment (magenta-af) [inline scrapy-webscanner/scanners/processors/pdf.py]: :+1:

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 6]
```diff
@@ -55,14 +55,14 @@ def convert(self, item, tmp_dir):
         command.extend(extra_options)
         command.append(new_file_path)
 
-        p = Popen(command, stdin=PIPE, stdout=DEVNULL, stderr=PIPE)
-        output, err = p.communicate(b"input data that is passed to subprocess' stdin")
+        completed_process = run(command, capture_output=True)
+        err = completed_process.stderr
 
-        if err:
+        if err is not None:
```
Comment (magenta-af) [inline scrapy-webscanner/scanners/processors/pdf.py]: Does this work properly? The docs suggest that `None` means "stream was not captured", not "stream was empty":
> Captured stderr from the child process. A bytes sequence, or a string if run() was called with an encoding, errors, or text=True. None if stderr was not captured.

**Comment by magenta-af** [comment]
I like the look of that now - thanks!

**Comment by dals83** [comment]
Okay I will see if the code actually run before I merge it :)

File: scrapy-webscanner/scanners/processors/pdf.py [hunk 7]
```diff
@@ -55,31 +55,26 @@ def convert(self, item, tmp_dir):
...
-            if err and 'Error' in err.decode('utf-8'):
-                print('pdftohtml convertion error: %s' % err)
-                return False
-        except TimeoutExpired as te:
-            p.kill()
-            # outs, errs = p.communicate()
-            print('Popen command expired: %s' % te)
-            return False
+        completed_process = run(command, stdout=PIPE, stderr=PIPE)
+        err = completed_process.stderr
```
Comment (danchr) [inline scrapy-webscanner/scanners/processors/pdf.py]: I'd just inline this variable; having it doesn't really make the code any more readable.

Comment (dals83) [reply scrapy-webscanner/scanners/processors/pdf.py]: The err variable?

**Pull request merged**